### PR TITLE
Fix saving XLSX with drawings

### LINKED ISSUE #544: Reusing a saved file which contained an image crashes
This is:

```
- [x] a bug report
- [ ] a feature request
- [ ] **not** a usage question (ask them on https://stackoverflow.com/questions/tagged/phpspreadsheet or https://gitter.im/PHPOffice/PhpSpreadsheet)
```

### What is the expected behavior?
A spreadsheet should be saved and opened multiple times, even if it contains images.

### What is the current behavior?
When opening and saving the file again, after it was saved, the following PHP error is generated:
```
Fatal error: Uncaught PhpOffice\PhpSpreadsheet\Writer\Exception: 
File zip://test-out.xlsx#xl/media/image11.png does not exist in PhpOffice/PhpSpreadsheet/Writer/Xlsx/ContentTypes.php:203
Stack trace: 
#0 PhpOffice/PhpSpreadsheet/Writer/Xlsx/ContentTypes.php(129): PhpOffice\PhpSpreadsheet\Writer\Xlsx\ContentTypes->getImageMimeType('zip://test-ou...') 
#1 PhpOffice/PhpSpreadsheet/Writer/Xlsx.php(223): PhpOffice\PhpSpreadsheet\Writer\Xlsx\ContentTypes->writeContentTypes(Object(PhpOffice\PhpSpreadsheet\Spreadsheet), false) 
#2 test.php(119): PhpOffice\PhpSpreadsheet\Writer\Xlsx->save('test-out.xlsx') 
#3 {main} thrown in PhpOffice/PhpSpreadsheet/Writer/Xlsx/ContentTypes.php on line 203
```

### What are the steps to reproduce?
Start a new document with Excel 2016 insert a small PNG image and save the document.
Upload it to webserver.

Please provide a [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve) of code that exhibits the issue without relying on an external Excel file or a web server:

```php
<?php

require __DIR__ . '/vendor/autoload.php';

// Create new Spreadsheet object
$spreadsheet = IOFactory::load('test-in.xls');
$writer = IOFactory::createWriter($spreadsheet, 'Xlsx');
$writer->save('test-out.xlsx');

unset($spreadsheet, $writer); // Or a page reload

$spreadsheet = IOFactory::load('test-out.xls');

// add code that show the issue here...
$writer = IOFactory::createWriter($spreadsheet, 'Xlsx');
$writer->save('test-out.xlsx');
```

### Which versions of PhpSpreadsheet and PHP are affected?
Just downloaded the lastest version here.
PHP version 7.0.23

### LINKED ISSUE #610: Inserting a drawing / chart causes print options to break in version 1.3.1
This is:

```
- [x] a bug report
- [ ] a feature request
- [ ] **not** a usage question (ask them on https://stackoverflow.com/questions/tagged/phpspreadsheet or https://gitter.im/PHPOffice/PhpSpreadsheet)
```

### What is the expected behavior?
In version 1.2.1 there was no issue with print options: I don't set any and the template I use is issue checked so that there shouldn't be any problems.

### What is the current behavior?
In version 1.3.1 I get an issue with Print Options. Excel repairs the file and removes print options.

### What are the steps to reproduce?

Please provide a [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve) of code that exhibits the issue without relying on an external Excel file or a web server:

```php
<?php

require __DIR__ . '/vendor/autoload.php';

// Create new Spreadsheet object
$spreadsheet = new \PhpOffice\PhpSpreadsheet\Spreadsheet();

// add code that show the issue here...
```

### LINKED ISSUE #676: Some images get lost after reading and writing xlsx-template file
This is:

```
- [x ] a bug report
- [ ] a feature request
- [ ] **not** a usage question (ask them on https://stackoverflow.com/questions/tagged/phpspreadsheet or https://gitter.im/PHPOffice/PhpSpreadsheet)
```

### What is the expected behavior?
When opening a template (which contains images), adding some text in an empty cell and saving the file, the images which are already contained in the template should not get deleted.

### What is the current behavior?
When opening a template (which contains images), adding some text in an empty cell and saving the file, SOME images (not all) and with-excel-created lines/Graphics (e.g. arrows) which are already contained in the template get deleted.

### What are the steps to reproduce?

Please provide a [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve) of code that exhibits the issue without relying on an external Excel file or a web server:

```php
   require 'vendor/autoload.php';
   use PhpOffice\PhpSpreadsheet\IOFactory;
   use PhpOffice\PhpSpreadsheet\Spreadsheet;
   use PhpOffice\PhpSpreadsheet\Writer\Xlsx;

   $reader = IOFactory::createReader('Xlsx');

   $inputFileName="mypath/myfile.xlsx";
   $outputFileName=$inputFileName;
   
   //Reading the XLS-File, which contains some images
   $spreadsheet = $reader->load($inputFileName);
   $sheet = $spreadsheet->getActiveSheet();

   // Write Text in Cell
   $sheet->setCellValue('B28', "Thats a test");

  // Overwrite the file (which is made here pseudo by deleting the original file and writing the new file with same name)
  // Save new file with new name
  $writer = new Xlsx($spreadsheet);
  $writer->save($outputFileName."_");
   // 2. Delete original file
  unlink($outputFileName);
  // Rename new file to original-filename
  rename($outputFileName."_", $outputFileName);
```

### Which versions of PhpSpreadsheet and PHP are affected?
PHP-Version: 7.2
PHPSpreadsheet-Version: April 2018 (where can i get the version number?)

### LINKED ISSUE #723: Corrupt file since migrating to 1.3.0 for complex doc when using removeSheet
This is:

```
- [x] a bug report
- [ ] a feature request
- [ ] **not** a usage question (ask them on https://stackoverflow.com/questions/tagged/phpspreadsheet or https://gitter.im/PHPOffice/PhpSpreadsheet)
```

### What is the expected behavior?
Open the result .xlsx with Excel without warning (after the document was saved with PhpSpreadsheet)

[Classeur_Empty.xlsx](https://github.com/PHPOffice/PhpSpreadsheet/files/2488240/Classeur_Empty.xlsx)
[Classeur_Empty.xlsxOUT.xlsx](https://github.com/PHPOffice/PhpSpreadsheet/files/2488241/Classeur_Empty.xlsxOUT.xlsx)

### What is the current behavior?
Open the result .xlsx with Excel ask a confirm notice to open :

Ask :
> Microsoft Excel: Sorry ... We found a problem in the contents of "Class_Box - Copy.xlsxOUT.xlsx"
> Microsoft Excel : Désolé... Nous avons trouvé un problème dans le contenu de "Classeur_Model - Copie.xlsxOUT.xlsx"

After confirm :
> Excel has completed file level validation and repair. Some parts of the workbook may have been repaired or abandoned.
> Deleted Part: Print Options.
> Deleted part: Drawing form.
> Excel a terminé la validation et la réparation de niveau fichier. Certaines parties du classeur peuvent avoir été réparées ou abandonnées.
> Partie supprimée: Options d'impression.
> Partie supprimée: Forme de dessin.

[Classeur_Model.xlsx](https://github.com/PHPOffice/PhpSpreadsheet/files/2488242/Classeur_Model.xlsx)
[Classeur_Model.xlsxOUT.xlsx](https://github.com/PHPOffice/PhpSpreadsheet/files/2488239/Classeur_Model.xlsxOUT.xlsx)

### What are the steps to reproduce?

  - Updating phpoffice/phpspreadsheet (1.0.0 => 1.2.1)
OK : no problem

  - Updating phpoffice/phpspreadsheet (1.2.1 => 1.3.0)
KO : the problem appears

```php
<?php

require __DIR__ . '/vendor/autoload.php';

use PhpOffice\PhpSpreadsheet\Spreadsheet;
use PhpOffice\PhpSpreadsheet\IOFactory;
use PhpOffice\PhpSpreadsheet\Worksheet\Worksheet;
use PhpOffice\PhpSpreadsheet\Writer\Xlsx;

// Create new Spreadsheet object
$spreadsheet = new \PhpOffice\PhpSpreadsheet\Spreadsheet();

// $model = __DIR__ . DIRECTORY_SEPARATOR . 'Classeur_Empty.xlsx';
$model = __DIR__ . DIRECTORY_SEPARATOR . 'Classeur_Model.xlsx';

$fileOut = $model . 'OUT.xlsx';

$spreadsheet = IOFactory::load($model);

$worksheet = $spreadsheet->setActiveSheetIndex(2);
$worksheet->setTitle('Mesures Avant');

$worksheet = $spreadsheet->setActiveSheetIndexByName('Mesures Avant');
$sheetIndex = $spreadsheet->getIndex($spreadsheet->getSheetByName('Mesures Avant'));
$spreadsheet->removeSheetByIndex($sheetIndex);	// without this removing, the output document is clean

$writer = IOFactory::createWriter($spreadsheet, 'Xlsx');
$writer->save($fileOut);

// now try to open "Classeur_Model - Copie.xlsxOUT.xlsx" with Excel
```

### Which versions of PhpSpreadsheet and PHP are affected?
All PhpSpreadsheet versions since 1.3.0 are affected.
(Used PHP Version 7.1.22)

### LINKED ISSUE #795: Wrong Ids relations when sheet clone
This is:

```
- [x] a bug report
- [ ] a feature request
- [ ] **not** a usage question (ask them on https://stackoverflow.com/questions/tagged/phpspreadsheet or https://gitter.im/PHPOffice/PhpSpreadsheet)
```

### What is the expected behavior?
Correct saved xlsx file.

### What is the current behavior?
File is broken.

### What are the steps to reproduce?
- I have xlsx document, contains one empty page with only one image (example file attached);
- load xlsx;
- clone first sheet and add to spreadsheet;
- save to a file.

```php
<?php
$template = IOFactory::load('some-src.xlsx');
$sheet = $template->getSheet(0)->copy();
$sheet->setTitle('Some sheet 2');
$template->addSheet($sheet);
$writer = new Xlsx($template);
$writer->save('some-dst.xlsx');
```

So... I unzip some-dst.xlsx and see file "xl/worksheets/_rels/sheet2.xml.rels":
<Relationship Id="rId1" Type=".../drawing" Target="../drawings/drawing2.xml"/>
<Relationship Id="rId1" Type=".../printerSettings" Target="../printerSettings/printerSettings1.bin"/>

There two relations have some Id (rId1).
Please, looks at the PhpOffice\PhpSpreadsheet\Writer\Xlsx\Rels::writeWorksheetRelationships(). I think that array $drawingOriginalIds have to contains items for new clonned page. I think there need to update array $drawingOriginalIds when any sheet will be cloned.

To fix this a use this way, but there is a very bad solution...
```php
<?php
$template = IOFactory::load('some-src.xlsx');
$sheet = $template->getSheet(0)->copy();
$sheet->setTitle('Some sheet 2');
$template->addSheet($sheet);

$raw = $template->getUnparsedLoadedData();
$drawingOriginalIds = $raw['sheets'][$sheet->getCodeName()]['drawingOriginalIds'];
$cnt = count($drawingOriginalIds);
$key = '../drawings/drawing' . $cnt . '.xml';
$drawingOriginalIds['../drawings/drawing' . ($cnt + 1) . '.xml'] = $drawingOriginalIds[$key];
$raw['sheets'][$sheet->getCodeName()]['drawingOriginalIds'] = $drawingOriginalIds;
$template->setUnparsedLoadedData($raw);

$writer = new Xlsx($template);
$writer->save('some-dst.xlsx');
```

### Which versions of PhpSpreadsheet and PHP are affected?
PHP 7.1
Spreadsheet 1.5.2

### LINKED ISSUE #853: PhpSpreadsheet can't load xlsx files created by itself
This is:

```
- [X] a bug report
```

The package is unable to load xlsx files created by itself with fatal error. There is no issue loading files created by Google Sheets (I don't have MS Office installed). The test.xlsx file was created by Google Sheets. 

### Test script

```php
<?php
// test.php
include 'vendor/autoload.php';
$spreadsheet = \PhpOffice\PhpSpreadsheet\IOFactory::load('test.xlsx');
$writer = \PhpOffice\PhpSpreadsheet\IOFactory::createWriter($spreadsheet, "Xlsx");
$writer->save('output.xlsx');
$spreadsheet = \PhpOffice\PhpSpreadsheet\IOFactory::load('output.xlsx');
```

### What is the expected behavior?
Excel file loaded into the $preadsheet variable

### What is the current behavior?
```
Fatal error: Uncaught Error: Call to a member function children() on bool in C:\...\vendor\phpoffice\phpspreadsheet\src\PhpSpreadsheet\Reader\Xlsx.php:1633
Stack trace:
#0 C:\...\vendor\phpoffice\phpspreadsheet\src\PhpSpreadsheet\IOFactory.php(95): PhpOffice\PhpSpreadsheet\Reader\Xlsx->load('out.xlsx')
#1 C:\...\test.php(7): PhpOffice\PhpSpreadsheet\IOFactory::load('out.xlsx')
#2 {main}
  thrown in C:\...\vendor\phpoffice\phpspreadsheet\src\PhpSpreadsheet\Reader\Xlsx.php on line 1633
```

### Which versions of PhpSpreadsheet and PHP are affected?
PHP 7.3.1
PhpSpreadsheet 1.6.0

Edit:
If I comment everything related to reading drawings everything works fine.

### LINKED ISSUE #861: Break xlsx file if you set image and print rule multiple sheet.
This is:

```
- [*] a bug report
```

### What is the expected behavior?
File saved normally.

### What is the current behavior?
If you save a file using a template that has multiple sheets and each sheet has a print setting and some images on some sheets, the file is broken. 

### What are the steps to reproduce?

Please provide a [Minimal, Complete, and Verifiable example](https://stackoverflow.com/help/mcve) of code that exhibits the issue without relying on an external Excel file or a web server:

1. Create a new template xlsx file.
2. Create multiple sheets in xlsx. (ex. create 4 sheets.)
3. Insert an image on each sheet. However, one sheet is set in the middle of the sheet, and a sheet with no image is set. (Do not save images on 3rd sheet)
4. On each sheet, set printer setting.Select "File" → "Print", and set printer (ex.Adobe PDF).
5. Save the above file in xlsx format.
6. Read the xlsx file you created and execute the following code with PhpSpreadSheet.

```php
<?php

require __DIR__ . '/vendor/autoload.php';

use PhpOffice\PhpSpreadsheet\IOFactory;
use PhpOffice\PhpSpreadsheet\Writer\Xlsx;

$path = "C:\\work\\Documents\\phpspreadsheet\\document.xlsx"; // the template file you created.
$reader = IOFactory::createReader('Xlsx');
$spreadsheet = $reader->load($path);

$writer = new Xlsx($spreadsheet);

$path_write = "C:\\work\\Documents\\phpspreadsheet\\output.xlsx"; // output file. this file is broken.
$writer->save($path_write);

```
If open output.xlsx, this is broken.
And recover this, the image of the next sheet of the sheet for which no image was set is lost.

You say "without relying on an external Excel file", but I found this bug only creating xlsx template file. So please create using this step.

### Which versions of PhpSpreadsheet and PHP are affected?
PhpSpreadsheet 1.6.0
PHP 7.2.4

### LINKED ISSUE #876: Incorred relation Id for existing drawings
This is:

```
- [X] a bug report
- [ ] a feature request
- [ ] **not** a usage question (ask them on https://stackoverflow.com/questions/tagged/phpspreadsheet or https://gitter.im/PHPOffice/PhpSpreadsheet)
```

### What is the expected behavior?
Open a file with a drawing in the 3rd sheet. Save it. Open it to view the content

### What is the current behavior?
Open it with excel raises a file corrupted error message
Open it with PHPSpreadsheet raises an error "Undefined index rId2"

### What are the steps to reproduce?
Here is the analysis
In the original file, the drawing is described by drawing1.xml and has "rId2" has relationship ID
Those data are stored in the $parent->unparsedLoadedData table

When saving, drawing1.xml is renamed drawing3.xml in the ZIP archive, but $unparsedLoadedData is not updated

So when the writer writes the drawing relations (Writer/Xlsx/Rels.php line 214), it looks in $unparsedLoadedData for drawing3.xml, does not find it and uses rId1 as relation ID instead of rId2 that is the original and actual one

I could submit a pull request to fix this, but I don't know the philosophy about handling both unparsedData and data that could be created/added with the library. If I simply update $unparsedLoadedData, I fear to create conflicts between unparsed content and managed content

Let me know how it should work

### Which versions of PhpSpreadsheet and PHP are affected?
1.6.0

### LINKED ISSUE #1022: Corrupt XLSX file when drawing image
This is:

```
- [* ] a bug report
- [ ] a feature request
- [ ] **not** a usage question (ask them on https://stackoverflow.com/questions/tagged/phpspreadsheet or https://gitter.im/PHPOffice/PhpSpreadsheet)
```
I wanted to download xlsx file in which added image file with phpspreadsheet.
Then I am facing issue of excel file broken due to "$drawing" as below source code.

### What is the expected behavior?
No error when open the excel file.

### What is the current behavior?
When opened the xlsx file, it always asked if fix to see. 

### What are the steps to reproduce?
Kindly try below code...

// add code that show the issue here...

use PhpOffice\PhpSpreadsheet\Settings;
use PhpOffice\PhpSpreadsheet\Reader\Xlsx as XlsxReader;
use PhpOffice\PhpSpreadsheet\Writer\Xlsx as XlsxWriter;
use PhpOffice\PhpSpreadsheet\Writer\CSV as CSVWriter;

class TestsController extends AppController{

    public function output(){

        //read template
        $reader = new XlsxReader();
        $spreadsheet = $reader->load(WWW_ROOT.'files/template/empty.xlsx');
        $spreadsheet->setActiveSheetIndex(0);
        $sheet = $spreadsheet->getActiveSheet();

        //add image in excel
        $drawing = new \PhpOffice\PhpSpreadsheet\Worksheet\Drawing();
        $drawing->setName('Logo');
        $drawing->setDescription('Logo');
        $drawing->setPath(WWW_ROOT."docs/C20101Docs/20190404202343.png"); 
        $drawing->setCoordinates('A1');
        $drawing->setWorksheet($sheet);
        unset($drawing);

        //create tmp file
        $tmp_path = WWW_ROOT.'files/test_tmp.xlsx';
        $writer = new XlsxWriter($spreadsheet);
        ob_end_clean();
        $writer->save($tmp_path);

        $content = 'attachment;';
        $content .= 'filename*=UTF-8\'\''.rawurlencode("TEST.xlsx"); 
        $this->response->header('Content-Disposition', $content);
        $this->response->type('application/vnd.openxmlformats-officedocument.spreadsheetml.sheet');
        $this->response->file($tmp_path);

        return $this->response;
    }
}

### Which versions of PhpSpreadsheet and PHP are affected?
PhpSpreadsheet 1.7
PHP 7.0.3
cakephp3.6

### LINKED ISSUE #1222: Opening and closing file corupts image
When opening an Excel file then writing to a new file the image stamped to each worksheet is corrupted.
### What is the expected behavior?
The image should remain

### What is the current behavior?
The image becomes corrupted

### What are the steps to reproduce?
```php 
$reader = IOFactory::createReaderForFile($path);
$excel = $reader->load($path);
$fileName = "test";
$writer = IOFactory::createWriter($excel,"Xlsx");
$writer->save($fileName.".xlsx");
```

### Which versions of PhpSpreadsheet and PHP are affected?
PHP 7.3.7 PHPSpreadsheet ^1.9

### LINKED ISSUE #1236: Xlsx writer - wrong rId for Drawings in Rels, when printerSettings is present in file
This is:

```
- [X ] a bug report
- [ ] a feature request
- [ ] **not** a usage question (ask them on https://stackoverflow.com/questions/tagged/phpspreadsheet or https://gitter.im/PHPOffice/PhpSpreadsheet)
```

### What is the expected behavior?
After addSheet(), which contains printer settings and images at the same time, images are displayed correctly (have correct rId in xl/worksheets/sheet2.xml.rels)

### What is the current behavior?
After addSheet(), which contains printer settings and images at the same time, file is opened as corupted and images are discarded - removed from file.

rId in xl/worksheets/sheet2.xml.rels has wrong index, because number one is taken by printerSettings

### What are the steps to reproduce?
addSheet(), which contains printer settings and images at the same time - the sheet is created in MS Office

### Which versions of PhpSpreadsheet and PHP are affected?
1.9.0 and previous versions

This worked for me
PROPOSED PATCH:
/phpoffice/phpspreadsheet/src/PhpSpreadsheet/Writer/Xlsx/Rels.php
line: 200

ADD:
if (isset($unparsedLoadedData['sheets'][$pWorksheet->getCodeName()]['printerSettings'])&&is_array($unparsedLoadedData['sheets'][$pWorksheet->getCodeName()]['printerSettings'])) {
          $d += count($unparsedLoadedData['sheets'][$pWorksheet->getCodeName()]['printerSettings']);
        }

EXPLANATION:
If unparsed data contatains 'printerSettings' rise referenceID

NOT SURE if printerSettings have always referenceID=1, so the patch might not be universal

### LINKED ISSUE #1294: Drawings get lost addExternalSheet
This is:

```
- [x] a bug report
```

### What is the expected behavior?

two xlsx-template files with one sheet and a image should combined into one xlsx with two sheets.

### What is the current behavior?

xlsx-files gets combined and corrupted, recovered xlsx removes the image on second sheet.

maybe its the same cause: see #1140, #927, #1022 https://github.com/PHPOffice/PhpSpreadsheet/issues/877#issuecomment-515455684 

### What are the steps to reproduce?

Please provide a [Minimal, Complete, and Verifiable example](https://stackoverflow.com/help/mcve) of code that exhibits the issue without relying on an external Excel file or a web server:

```php
<?php

require __DIR__ . '/vendor/autoload.php';

$reader = new \PhpOffice\PhpSpreadsheet\Reader\Xlsx ();
$spfirst = $reader->load ( "first.xlsx" );
$spsecond = $reader->load ( "second.xlsx" );
$pSheet = $spsecond->getActiveSheet ();
$spfirst->addExternalSheet ( $pSheet, 0 );

$objWriter = new \PhpOffice\PhpSpreadsheet\Writer\Xlsx ( $spfirst );
$objWriter->setPreCalculateFormulas ( true );

$objWriter->save ( "result.xlsx" );


Example files:

[first.xlsx](https://github.com/PHPOffice/PhpSpreadsheet/files/3991477/first.xlsx)
[second.xlsx](https://github.com/PHPOffice/PhpSpreadsheet/files/3991479/second.xlsx)

[result.xlsx](https://github.com/PHPOffice/PhpSpreadsheet/files/3991478/result.xlsx)

### Which versions of PhpSpreadsheet and PHP are affected?

1.10.1 / 1.8.2

### LINKED ISSUE #1421: The picture is not copied to a new sheet
This is:

```
- [ ] a bug report
- [ ] a feature request
- [x ] **not** a usage question (ask them on https://stackoverflow.com/questions/tagged/phpspreadsheet or https://gitter.im/PHPOffice/PhpSpreadsheet)
```

### What is the expected behavior?
new sheet contains images

### What is the current behavior?
new sheet does not contain a images

### What are the steps to reproduce?

xls
[test.xlsx](https://github.com/PHPOffice/PhpSpreadsheet/files/4361824/test.xlsx)

```php
<?php

require __DIR__ . '/vendor/autoload.php';

use PhpOffice\PhpSpreadsheet\Spreadsheet;
use PhpOffice\PhpSpreadsheet\Writer\Xlsx;
$outfile ='new_doc.xlsx';
$spreadsheet = \PhpOffice\PhpSpreadsheet\IOFactory::load('test.xlsx');
$spreadsheet->setActiveSheetIndex(1);

$drawing = new \PhpOffice\PhpSpreadsheet\Worksheet\Drawing();
$drawing->setName('img');
$drawing->setDescription('img');
$drawing->setPath('xls.png');
$drawing->setCoordinates('B8');
$drawing->setWorksheet($spreadsheet->getActiveSheet());

$newSheet = $spreadsheet->getActiveSheet()->copy();
$newSheet->setTitle("new_sheet");
$spreadsheet->addSheet($newSheet, 2);
$spreadsheet->setActiveSheetIndex(2);

$writer = new Xlsx($spreadsheet);
$writer->save($outfile);
```

### Which versions of PhpSpreadsheet and PHP are affected?
php 7.3.1,  PhpSpreadsheet 1.11.0

### PR DESCRIPTION
**Pull request by Vagir-dev**
This is:

```
- [x] a bugfix
- [ ] a new feature
```

Checklist:

- [x] Changes are covered by unit tests
- [x] Code style is respected
- [x] Commit message explains why the change is made
- [x] CHANGELOG.md contains a short summary of the change
- [ ] Documentation is updated as necessary

### Why this change is needed?

A problem with broken files is often happens when PhpSpreadsheet writes XLSX file format.
Opening such files in MS Office raises error messages like

> We found a problem with some content in filename.xlsx.
> Do you want us to try and recover as much as we can? If you trust the source of this workbook, click Yes.

And then after going through the repair process we get the message from Excel that some of images (or sometimes other elements) were deleted from the spreadsheet.

This problem happens when any XLSX file was loaded as a template and there is at least one drawing in it.

The reason is in generating incorrect `rId`s for `worksheets\_rels`. Here is example (doubled **rId1**):

_xl\worksheets\_rels\sheet1.xml.rels_
```
<?xml version="1.0" encoding="UTF-8" standalone="yes" ?>
- <Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships">
  <Relationship Id="rId1" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/drawing" Target="../drawings/drawing1.xml" />
  <Relationship Id="rId1" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/printerSettings" Target="../printerSettings/printerSettings1.bin" />
  </Relationships>
```

Analysing code of `PhpSpreadsheet\Writer\Xlsx\` shows that algorithm used for writing spreadsheet's content to XML files cannot generate `rId`s in right sequence.
It's because `rId`s for `worksheets\_rels` are generated separately for each rel Type (`printerSettings`, `drawing`, `vmlDrawings` etc.) but not as a continuous sequence.
The fact that XLSX files created by the PhpSpreadsheet from a template are often correct is a luck, coincidence - when `rId`s from the template and `rId`s from programmatically created elements suite each other.
I hope you understand what I mean...

By the way, it is why such problem doesn't arise if spreadsheet created totally by code, not by loading from template.

To fix this completely a big code refactoring is needed, so it is unreal for now, unfortunately.
But in many cases (not all unfortunately) we can avoid described problem by adding a smal changes in `Writer/Xlsx/Rels` class.

Hope this will help.

Test demonstrates one of cases: when 2nd (3rd, 4th...) sheet contains any drawing(s) but the first one doesn't.

Sorry for my pure English, it is not my native language.

Resolves #1421
Resolves #1294
Resolves #1236
Resolves #1222
Resolves #1022
Resolves #876
Resolves #861
Resolves #853
Resolves #795
Resolves #723
Resolves #676
Resolves #610
Resolves #544

### CODE REVIEW

**Comment by MAKS-dev** [comment]
Would be nice if #1153 would also be taken into consideration.

**Comment by Vagir-dev** [comment]
> Would be nice if #1153 would also be taken into consideration.

@MAKS-dev, I encountered this problem too.
I think I resolved it locally. Maybe I'll make a PR, when current PR will be approved.

**Comment by maosong-pactera** [comment]
@Vagir-dev I used your latest code, but there are still problems

clone worksheet

```php
<?php
use PhpOffice\PhpSpreadsheet\Spreadsheet;
use PhpOffice\PhpSpreadsheet\Writer\Xlsx;

$outfile ='new_doc.xlsx';
$spreadsheet = \PhpOffice\PhpSpreadsheet\IOFactory::load(__DIR__ . '/new.xlsx');
$clonedWorksheet = clone $spreadsheet->getSheetByName('Sheet1');
$clonedWorksheet->setTitle('Sheet2');
$spreadsheet->addSheet($clonedWorksheet);
$writer = new Xlsx($spreadsheet);
$writer->save($outfile);
```

**Comment by Vagir-dev** [comment]
> @Vagir-dev I used your latest code, but there are still problems
> 
> clone worksheet

@maosong-pactera, thank you for your reply.
Yes, you're right, cloning is not affected by my fix. This is what I mean writing 
> But in many cases (**not all** unfortunately) we can avoid described problem...
My fix is just a small step to avoid some part of problems.
Now I'll try to solve cloning problem too.
Thank you!

**Comment by Vagir-dev** [comment]
@maosong-pactera, Ican't repeat cloning problem with my template files. Can you please attach your file 'new.xlsx'?

> ```
> $spreadsheet = \PhpOffice\PhpSpreadsheet\IOFactory::load(__DIR__ . '/new.xlsx');
> ```

**Comment by Vagir-dev** [comment]
> I used your latest code, but there are still problems

@maosong-pactera, are you shure you applied my fix to the sources?

**Comment by Howitzer** [comment]
@Vagir-dev thank you so much for this! I was getting the rId errors and corrupt XLSX file only when I used cloning in my sheets, but applying your fix has made everything work now. You saved my day!

File: src/PhpSpreadsheet/Writer/Xlsx/Rels.php [hunk 1]
```diff
@@ -194,7 +194,6 @@ public function writeWorksheetRelationships(\PhpOffice\PhpSpreadsheet\Worksheet\
         $objWriter->writeAttribute('xmlns', 'http://schemas.openxmlformats.org/package/2006/relationships');
 
         // Write drawing relationships?
-        $d = 0;
```
Comment (Vagir-dev) [inline src/PhpSpreadsheet/Writer/Xlsx/Rels.php:197]: **$d** will be used only once - in the row 212

File: src/PhpSpreadsheet/Writer/Xlsx/Rels.php [hunk 2]
```diff
@@ -208,13 +207,18 @@ public function writeWorksheetRelationships(\PhpOffice\PhpSpreadsheet\Worksheet\
         }
 
         if (($pWorksheet->getDrawingCollection()->count() > 0) || (count($charts) > 0) || $drawingOriginalIds) {
-            $relPath = '../drawings/drawing' . $pWorksheetId . '.xml';
-            $rId = ++$d;
```
Comment (Vagir-dev) [inline src/PhpSpreadsheet/Writer/Xlsx/Rels.php:212]: so lets combine two lines into one

File: src/PhpSpreadsheet/Writer/Xlsx/Rels.php [hunk 3]
```diff
@@ -208,13 +207,18 @@ public function writeWorksheetRelationships(\PhpOffice\PhpSpreadsheet\Worksheet\
...
+            $rId = 1;
 
+            // Use original $relPath to get original $rId.
+            // take first. In future can be overriten
+            // (! synchronize with \PhpOffice\PhpSpreadsheet\Writer\Xlsx\Worksheet::writeDrawings)
```
Comment (Vagir-dev) [inline src/PhpSpreadsheet/Writer/Xlsx/Rels.php:214]: Comment took from
\PhpOffice\PhpSpreadsheet\Writer\Xlsx\Worksheet::writeDrawings
And an link to it added to show a correlation between this methods: they both use th same [temporary] way of getting $relPath from $drawingOriginalIds array.

File: src/PhpSpreadsheet/Writer/Xlsx/Rels.php [hunk 4]
```diff
@@ -208,13 +207,18 @@ public function writeWorksheetRelationships(\PhpOffice\PhpSpreadsheet\Worksheet\
...
+            $relPath = array_key_first($drawingOriginalIds);
```
Comment (Vagir-dev) [inline src/PhpSpreadsheet/Writer/Xlsx/Rels.php:215]: This is the most important line in this Fix!
Only such $relPath can be used properly in the next two lines.

File: src/PhpSpreadsheet/Writer/Xlsx/Rels.php [hunk 5]
```diff
@@ -208,13 +207,18 @@ public function writeWorksheetRelationships(\PhpOffice\PhpSpreadsheet\Worksheet\
         }
 
         if (($pWorksheet->getDrawingCollection()->count() > 0) || (count($charts) > 0) || $drawingOriginalIds) {
-            $relPath = '../drawings/drawing' . $pWorksheetId . '.xml';
```
Comment (Vagir-dev) [inline src/PhpSpreadsheet/Writer/Xlsx/Rels.php:211]: Just moved down into line 221.

File: src/PhpSpreadsheet/Writer/Xlsx/Rels.php (hunk 4 above)
Comment (CiTRO33) [reply src/PhpSpreadsheet/Writer/Xlsx/Rels.php:215]: array_key_first works only >= php 7.3 

https://www.php.net/manual/en/function.array-key-first.php

PhpSpreadsheet also supports 7.2 until Nov. 2020

Comment (Vagir-dev) [reply src/PhpSpreadsheet/Writer/Xlsx/Rels.php]: My mistake.
Fixed.
Thank you!

**Comment by Vagir-dev** [comment]
@PowerKiKi, hi!
May I ask why this PR is not merged?
Am I doing something wrong?

**Comment by MarkBaker** [comment]
Sorry that it's taken me a while to assess the code and confirm that it resolves the basic problem. As you say, some significant changes would be needed to provide a full resolution, but this certainly resolves the rid issues in many commonly-occurring cases
Thank you for the your time and effort

**Comment by Vagir-dev** [comment]
@MarkBaker, thank you!

**Comment by Aristotel2003** [comment]
Hello.  I have latest version PhpSpreadsheet [1.13.0] - 2020-05-31

**Comment by Vagir-dev** [comment]
> I get the same error.
> I am load excel from template
> and insert image.

Hi @Aristotel2003 

This is not the same case. My PR fixes cases when the template file contains at least one drawing.
But your template doesn't.
So I think you should start a new [issue](https://github.com/PHPOffice/PhpSpreadsheet/issues) for your problem.

**Comment by Aristotel2003** [comment]
I am great, but my fix will suit you too. He makes a correction in 1 line.

вт, 16 июн. 2020 г. в 23:26, Vagir <notifications@github.com>:

**Comment by Vagir-dev** [comment]
> I am great, but my fix will suit you too. He makes a correction in 1 line.

That's great! So why don't you want to create your own Pull Request?

**Pull request merged**